### ISSUE TITLE
Composer crashes when clearing cache

### ISSUE THREAD

**Issue by lfuelling** [roundcube/plugin-installer#30]
Hi,

I'm setting up roundcube using Ansible. This setup is a bit old but worked fine a few minutes ago until suddenly (I had to run the playbook again) it crashed (and now keeps crashing) at the "clear cache" step.

The step in question is the following:
```yaml
- name: Clear Composer cache
  shell: COMPOSER_ALLOW_SUPERUSER=1 php /root/composer.phar clearcache -n
           chdir=/var/www/roundcube
```

This should only clear the composer cache but it crashes with the following output:
```
Fatal error: Uncaught Error: Class \"Roundcube\\Composer\\PluginInstaller\" not found in phar:///root/composer.phar/src/Composer/Plugin/PluginManager.php:202
Stack trace:
#0 phar:///root/composer.phar/src/Composer/Plugin/PluginManager.php(267): Composer\\Plugin\\PluginManager->registerPackage()
#1 phar:///root/composer.phar/src/Composer/Plugin/PluginManager.php(77): Composer\\Plugin\\PluginManager->loadRepository()
#2 phar:///root/composer.phar/src/Composer/Factory.php(384): Composer\\Plugin\\PluginManager->loadInstalledPlugins()
#3 phar:///root/composer.phar/src/Composer/Factory.php(576): Composer\\Factory->createComposer()
#4 phar:///root/composer.phar/src/Composer/Console/Application.php(368): Composer\\Factory::create()
#5 phar:///root/composer.phar/src/Composer/Console/Application.php(486): Composer\\Console\\Application->getComposer()
#6 phar:///root/composer.phar/src/Composer/Console/Application.php(178): Composer\\Console\\Application->getPluginCommands()
#7 phar:///root/composer.phar/vendor/symfony/console/Application.php(117): Composer\\Console\\Application->doRun()
#8 phar:///root/composer.phar/src/Composer/Console/Application.php(113): Symfony\\Component\\Console\\Application->run()
#9 phar:///root/composer.phar/bin/composer(61): Composer\\Console\\Application->run()
#10 /root/composer.phar(24): require('...')
#11 {main}
thrown in phar:///root/composer.phar/src/Composer/Plugin/PluginManager.php on line 202
```

I'm currently using the latest 1.x version of composer because 2.x generates an additional deprecation warning. But I've tested this with latest 1.x and latest 2.x and both produce the same crash dump.

Here's the composer json I'm using:
```json
{
    "name": "roundcube/roundcubemail",
    "description": "The Roundcube Webmail suite",
    "license": "GPL-3.0-or-later",
    "repositories": [
        {
            "type": "composer",
            "url": "https://plugins.roundcube.net/"
        },
        {
            "type": "vcs",
            "url": "https://github.com/roundcube/Net_Sieve.git"
        },
        {
            "type": "vcs",
            "url": "https://git.kolab.org/diffusion/PNL/php-net_ldap.git"
        }
    ],
    "require": {
        "php": ">=5.5.0",
        "pear/pear-core-minimal": "~1.10.1",
        "pear/auth_sasl": "~1.1.0",
        "pear/mail_mime": "~1.10.0",
        "pear/net_smtp": "~1.9.0",
        "pear/net_socket": "~1.2.2",
        "pear/net_idna2": "~0.2.0",
        "pear/crypt_gpg": "~1.6.3",
        "pear/net_sieve": "~1.4.3",
        "masterminds/html5": "~2.7.0",
        "roundcube/net_sieve": "~1.5.0",
        "endroid/qr-code": "~1.6.5",
        "alexandregz/twofactor_gauthenticator": "dev-master",
	    "roundcube/carddav": "dev-master",
        "guzzlehttp/guzzle": "^6.5.5"
    },
    "require-dev": {
        "phpunit/phpunit": "^4.8.36 || ^5.7.21 || ^6 || ^7"
    },
    "suggest": {
        "pear/net_ldap2": "~2.2.0 required for connecting to LDAP address books",
        "bjeavons/zxcvbn-php": "^1.0 required for Zxcvbn password strength driver",
        "kolab/net_ldap3": "~1.1.1 required for connecting to LDAP"
    },
    "minimum-stability": "dev"
}
```

I'm pretty confused by this sudden crash because (as stated above) the setup worked fine a few minutes prior.

**EDIT:** I've read the comment in #16 suggesting to manually add the plugin-installer. That worked, but I'm still confused why this happened without any changes to the roundcube installation. The scripts download a fixed roundcube version (1.5-beta) and the logs show that this step was skipped (only the composer "reinstall" happens every run). The same playbook also installs a second roundcube instance with the same version, on the same host, using the same composer executable and this crash doesn't happen there.

**EDIT 2:** It seems that switching from `1.5-beta` to `1.5-rc` "fixes" the crash, cloning the plugin-installer manually is no longer required. (The other roundcube instance is still on `1.5-beta` and still works without issues.)

**Comment by adjobscc-del** [comment]
> Hi,
> 
> I'm setting up roundcube using Ansible. This setup is a bit old but worked fine a few minutes ago until suddenly (I had to run the playbook again) it crashed (and now keeps crashing) at the "clear cache" step.
> 
> The step in question is the following:
> 
> ```yaml
> - name: Clear Composer cache
>   shell: COMPOSER_ALLOW_SUPERUSER=1 php /root/composer.phar clearcache -n
>            chdir=/var/www/roundcube
> ```
> 
> This should only clear the composer cache but it crashes with the following output:
> 
> ```
> Fatal error: Uncaught Error: Class \"Roundcube\\Composer\\PluginInstaller\" not found in phar:///root/composer.phar/src/Composer/Plugin/PluginManager.php:202
> Stack trace:
> #0 phar:///root/composer.phar/src/Composer/Plugin/PluginManager.php(267): Composer\\Plugin\\PluginManager->registerPackage()
> #1 phar:///root/composer.phar/src/Composer/Plugin/PluginManager.php(77): Composer\\Plugin\\PluginManager->loadRepository()
> #2 phar:///root/composer.phar/src/Composer/Factory.php(384): Composer\\Plugin\\PluginManager->loadInstalledPlugins()
> #3 phar:///root/composer.phar/src/Composer/Factory.php(576): Composer\\Factory->createComposer()
> #4 phar:///root/composer.phar/src/Composer/Console/Application.php(368): Composer\\Factory::create()
> #5 phar:///root/composer.phar/src/Composer/Console/Application.php(486): Composer\\Console\\Application->getComposer()
> #6 phar:///root/composer.phar/src/Composer/Console/Application.php(178): Composer\\Console\\Application->getPluginCommands()
> #7 phar:///root/composer.phar/vendor/symfony/console/Application.php(117): Composer\\Console\\Application->doRun()
> #8 phar:///root/composer.phar/src/Composer/Console/Application.php(113): Symfony\\Component\\Console\\Application->run()
> #9 phar:///root/composer.phar/bin/composer(61): Composer\\Console\\Application->run()
> #10 /root/composer.phar(24): require('...')
> #11 {main}
> thrown in phar:///root/composer.phar/src/Composer/Plugin/PluginManager.php on line 202
> ```
> 
> I'm currently using the latest 1.x version of composer because 2.x generates an additional deprecation warning. But I've tested this with latest 1.x and latest 2.x and both produce the same crash dump.
> 
> Here's the composer json I'm using:
> 
> ```json
> {
>     "name": "roundcube/roundcubemail",
>     "description": "The Roundcube Webmail suite",
>     "license": "GPL-3.0-or-later",
>     "repositories": [
>         {
>             "type": "composer",
>             "url": "https://plugins.roundcube.net/"
>         },
>         {
>             "type": "vcs",
>             "url": "https://github.com/roundcube/Net_Sieve.git"
>         },
>         {
>             "type": "vcs",
>             "url": "https://git.kolab.org/diffusion/PNL/php-net_ldap.git"
>         }
>     ],
>     "require": {
>         "php": ">=5.5.0",
>         "pear/pear-core-minimal": "~1.10.1",
>         "pear/auth_sasl": "~1.1.0",
>         "pear/mail_mime": "~1.10.0",
>         "pear/net_smtp": "~1.9.0",
>         "pear/net_socket": "~1.2.2",
>         "pear/net_idna2": "~0.2.0",
>         "pear/crypt_gpg": "~1.6.3",
>         "pear/net_sieve": "~1.4.3",
>         "masterminds/html5": "~2.7.0",
>         "roundcube/net_sieve": "~1.5.0",
>         "endroid/qr-code": "~1.6.5",
>         "alexandregz/twofactor_gauthenticator": "dev-master",
> 	    "roundcube/carddav": "dev-master",
>         "guzzlehttp/guzzle": "^6.5.5"
>     },
>     "require-dev": {
>         "phpunit/phpunit": "^4.8.36 || ^5.7.21 || ^6 || ^7"
>     },
>     "suggest": {
>         "pear/net_ldap2": "~2.2.0 required for connecting to LDAP address books",
>         "bjeavons/zxcvbn-php": "^1.0 required for Zxcvbn password strength driver",
>         "kolab/net_ldap3": "~1.1.1 required for connecting to LDAP"
>     },
>     "minimum-stability": "dev"
> }
> ```
> 
> I'm pretty confused by this sudden crash because (as stated above) the setup worked fine a few minutes prior.
> 
> **EDIT:** I've read the comment in #16 suggesting to manually add the plugin-installer. That worked, but I'm still confused why this happened without any changes to the roundcube installation. The scripts download a fixed roundcube version (1.5-beta) and the logs show that this step was skipped (only the composer "reinstall" happens every run). The same playbook also installs a second roundcube instance with the same version, on the same host, using the same composer executable and this crash doesn't happen there.
> 
> **EDIT 2:** It seems that switching from `1.5-beta` to `1.5-rc` "fixes" the crash, cloning the plugin-installer manually is no longer required. (The other roundcube instance is still on `1.5-beta` and still works without issues.)

**Comment by mvorisek** [comment]
@alecpl let's close this issue as it seems fixed, OP can always comment later if not.
